possible compilation error when "using namespace std;" is defined before including "hpx" headers files

a compilation error may occur when "using namespace std;" is defined before including hpx header files. I can only reproduce this error with my code though and have no very simple test code where the error appears. 

## Expected Behavior

I expected the compilation not to fail.

## Actual Behavior

g++ returns this 

```
hpx/runtime/components/memory_block.hpp: In member function ‘hpx::components::access_memory_block_proxy<T>& hpx::components::access_memory_block_proxy<T>::operator=(const target_type&)’:
include/hpx/runtime/components/memory_block.hpp:185:20: error: invalid use of ‘class std::set<typename std::remove_const<_Tp>::type>’
             block_.template set<target_type>(rhs);
```
if we have this in an header file:

```
using namespace std;
#include <hpx/config.hpp>
#include <hpx/hpx.hpp>
```
removing the using namespace std; works fine. John suggested that it could be a clash between the std::set function and the set function defined in the class. the compiler try to resolve std::set before searching the set function in the class.

## Steps to Reproduce the Problem

Difficult to reproduce with a very simple code. Seems to be related to my code.

## Specifications

  - HPX Version: 1.2 (git from 04/13)
  - Platform (compiler, OS): power 8 linux 3.10.0 
  - gcc : 6.4.0
  - boost : 1.66.0

Mathieu

--- a/hpx/runtime/components/memory_block.hpp
+++ b/hpx/runtime/components/memory_block.hpp
@@ -182,13 +182,13 @@ namespace hpx { namespace components
 
         access_memory_block_proxy& operator=(target_type const& rhs)
         {
-            block_.template set<target_type>(rhs);
+            block_.template set_data<target_type>(rhs);
             return *this;
         }
 
         operator target_type const&() const
         {
-            return block_.template get<target_type>();
+            return block_.template get_data<target_type>();
         }
 
     private:

--- a/hpx/runtime/components/server/memory_block.hpp
+++ b/hpx/runtime/components/server/memory_block.hpp
@@ -246,19 +246,19 @@ namespace hpx { namespace components
         }
 
         template <typename T>
-        T& get()
+        T& get_data()
         {
             return *reinterpret_cast<T*>(get_ptr());
         }
 
         template <typename T>
-        T const& get() const
+        T const& get_data() const
         {
             return *reinterpret_cast<T const*>(get_ptr());
         }
 
         template <typename T>
-        void set (T const& val)
+        void set_data (T const& val)
         {
             if (!data_) {
                 std::ostringstream strm;
@@ -266,7 +266,7 @@ namespace hpx { namespace components
                      << components::get_component_type_name(component_memory_block)
                      << ")";
                 HPX_THROW_EXCEPTION(invalid_status,
-                    "memory_block_data::set",
+                    "memory_block_data::set_data",
                     strm.str());
             }
             if (!data_->is_master())
@@ -276,7 +276,7 @@ namespace hpx { namespace components
                      << components::get_component_type_name(component_memory_block)
                      << ")";
                 HPX_THROW_EXCEPTION(invalid_status,
-                    "memory_block_data::set",
+                    "memory_block_data::set_data",
                     strm.str());
             }
             *reinterpret_cast<T*>(data_->get_ptr()) = val;

